consensus/consensus-types/src/common.rs

```diff
@@ -139,14 +139,61 @@ impl ProofWithData {
     }
 }
 
+#[derive(Deserialize, Serialize, Clone, Debug)]
+pub struct ProofWithDataV2 {
+    pub proof_with_data: ProofWithData,
+    pub txns_to_include: Option<usize>,
+}
+
+impl PartialEq for ProofWithDataV2 {
+    fn eq(&self, other: &Self) -> bool {
+        self.proof_with_data == other.proof_with_data
+            && self.txns_to_include == other.txns_to_include
+    }
+}
+
+impl Eq for ProofWithDataV2 {}
+
+impl ProofWithDataV2 {
+    pub fn new(proof_with_data: ProofWithData, txns_to_include: Option<usize>) -> Self {
+        Self {
+            proof_with_data,
+            txns_to_include,
+        }
+    }
+
+    pub fn extend(&mut self, other: ProofWithDataV2) {
+        self.proof_with_data.extend(other.proof_with_data);
+        // InQuorumStoreV2 TODO: what is the right logic here ???
+        if self.txns_to_include.is_none() {
+            self.txns_to_include = other.txns_to_include;
+        }
+    }
+}
+
 /// The payload in block.
 #[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
 pub enum Payload {
     DirectMempool(Vec<SignedTransaction>),
     InQuorumStore(ProofWithData),
+    InQuorumStoreV2(ProofWithDataV2),
```

Comment:
thinking about it , maybe `InQuorumStoreWithLimit` is nicer name, and not too long?

Reply:
Yeah sounds good